File: frontend/lib/api/contacts.ts
/** Browser-side calls to the networking contacts API (`/api/v1/contacts`). */

import { requestJson, type ApiFailure } from "@/lib/api/request";

const CONTACTS_API = "/api/v1/contacts";

/** A `company_networking` row; `name` joins the first and last names (`null` when both are). */
export type Contact = {
  id: number;
  company_id: number;
  name: string | null;
  first_name: string | null;
  last_name: string | null;
  title: string | null;
  linkedin_url: string | null;
  connection_request_sent: boolean;
  /** ISO 8601 timestamp of the first time the request was marked as sent. */
  connection_request_sent_at: string | null;
};

export type ContactResult = { ok: true; contact: Contact } | ApiFailure;

/** Record that a connection request was sent; calling it again keeps the first timestamp. */
export async function markConnectionRequestSent(
  contactId: number,
): Promise<ContactResult> {
  const result = await requestJson<Contact>(
    `${CONTACTS_API}/${contactId}/connection-request`,
    { method: "POST" },
  );
  return result.ok ? { ok: true, contact: result.data } : result;
}

File: backend/jobs/src/api/v1/contacts.py
"""Networking contacts at a company (``/api/v1/contacts``)."""

import logging
from datetime import UTC, datetime

from fastapi import APIRouter, HTTPException, status

from src.api.v1.common import DbSession
from src.api.v1.schemas.contacts import ContactRead
from src.models import CompanyNetworking

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/contacts", tags=["contacts"])

CONTACT_NOT_FOUND = "Contact not found"


@router.post("/{contact_id}/connection-request")
def mark_connection_request_sent(contact_id: int, session: DbSession) -> ContactRead:
    """Record that a connection request was sent; calling it again keeps the first timestamp."""
    contact = session.get(CompanyNetworking, contact_id)
    if contact is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, CONTACT_NOT_FOUND)
    contact.connection_request_sent = True
    if contact.connection_request_sent_at is None:
        contact.connection_request_sent_at = datetime.now(UTC)
    session.commit()
    logger.info("Contact %d marked as connection request sent", contact_id)
    return ContactRead.from_model(contact)
